### DISCUSSION TITLE
Zod and optional number inputs

### DISCUSSION THREAD

**Discussion by mtnowl** [react-hook-form/react-hook-form#6980] [Q&A] [answered]
I'm trying to figure out how to get optional number inputs to work with Zod and react-hook-form. On the resolvers README, we have this:
```
const schema = z.object({
  name: z.string().nonempty({ message: 'Required' }),
  age: z.number().min(10),
});


  <input type="number" {...register('age', { valueAsNumber: true })} />
```

When validating if the input is empty, I get the "Expected number, received nan" error. This occurs regardless of if I modify the schema to have `.optional()` or not.

I even tried adding in `NaN` as a literal through a union `z.union([z.literal(NaN), z.number()])`, but then it gives me "Expected NaN, received NaN"

What is the intended way to handle optional numeric inputs? How do I interact with that error?

CodeSandbox: https://codesandbox.io/s/react-hook-form-number-input-error-c0cbs?file=/src/App.tsx

**Comment by bluebill1049** [comment]
share a codesandbox.

**Reply by mtnowl** [reply]
CodeSandbox added, link is https://codesandbox.io/s/react-hook-form-number-input-error-c0cbs?file=/src/App.tsx

**Comment by Sti2nd** [comment, marked as answer]
The reason you get `NaN` @henryjin3 is because `valueAsNumber` returns NaN if something goes wrong (from the [register documentation](https://react-hook-form.com/api/useform/register)). And something goes wrong because the input field is returning a string, something I learned in https://github.com/react-hook-form/react-hook-form/discussions/7128.

There is an option on register called `setValueAs` which let us define our own transform function that will be called before validation.

```ts
z.object({
  age: z.number().optional(),
});

<input
  type="number"
  {...register("age", {
    setValueAs: (v) => v === "" ? undefined : parseInt(v, 10),
  })}
/>;
```

**Reply by bryanprimus** [reply]
is this still the optimal solution? since we need to add a lot of code only for that functionality

**Reply by itujono** [reply]
Might be late to the party. But this'll do the trick without touching your base input component at all:
```js
z.object({
   age: z.union([z.number().int().positive().min(1), z.nan()]).optional(),
})
```

**Reply by aidanCQ** [reply]
I ended up modifying `z.nan()` to become `z.nan().transform(() => undefined)` to avoid having to handle the NaN case.

**Reply by robert-ovens** [reply]
@itujono you just made my day :-)

**Reply by Sustavo** [reply]
You're my heroes!

**Reply by backstage2000** [reply]
@itujono Nice solution!

**Comment by Claudius10** [comment]
Hi. I used coerce together with optional() and it works
z.object({
  aNumber: z.coerce.number().optional(),
});

https://zod.dev/?id=coercion-for-primitives

**Reply by coopbri** [reply]
I'm not sure about unintended side effects, but using coercion (`z.coerce`) worked for me as well with a complex string input. Thanks @Claudius10!

**Reply by adamsuskin** [reply]
Careful @coopbri. One edge case I found was that `null` gets coerced to `"null"` which at least for me is certainly unintended.

**Reply by coopbri** [reply]
Thanks a lot @adamsuskin, I didn't catch that and that's certainly good to know.

**Reply by luiscarlin** [reply]
For me empty string `""` (input cleared) was getting coerced to `0`, which is not good when I also have a `.min(1)` and `.optional()`

**Reply by kellyrmilligan** [reply]
what did you end up putting for your default values in this case?  I was trying this, and it wants a number, but it's just a string :)

**Reply by kellyrmilligan** [reply]
```
.object({
    name: z.string().trim().min(1, { message: "Name is required" }),
    nickname: z.string(),
    preferredSalutation: z.string(),
    budget: z.coerce.number(),
  })

export type Schema = z.infer<typeof schema>


const defaultValues: Schema = {
    name: "",
    nickname: "",
    preferredSalutation: "",
    budget: "", // ts error, wants a number
  }

```

**Reply by hatchli** [reply]
Did you get an answer for this? I'm in the same position

**Reply by its-monotype** [reply]
add `.or(z.literal('')))`

**Reply by henokyehulu** [reply]
What if I want it be required? Because your option make it that you can just send an empty string rather than a number. Please help!

**Reply by enfroz** [reply]
@henokyehulu  Try this:

https://github.com/colinhacks/zod/issues/2461#issuecomment-1712769467

**Comment by BaziotaBeans** [comment]
This work for me:

```ts 
suits: z.preprocess((value) => {
      if (value === "") return undefined;
      return Number(value);
    }, z.union([z.number().int().positive().min(1), z.nan()]).optional()),
```

**Comment by sibananda485** [comment]
Has anyone found a solid solution for this?

**Reply by max-wel** [reply]
Worked for me
```js
age: z
    .number({ message: "Age is required" })
    .refine((val) => !isNaN(val), { message: "Age is required" })
    .int()
    .positive()
    .min(18, { message: "Age should be at least 18" }),
